lib/google/cloud/env/lazy_value.rb

```diff
@@ -0,0 +1,1008 @@
+        # are ignored if a cached value is returned.
+        #
+        # @return [Object] the value
+        # @raise [Exception] if an error happened while computing the value
+        #
+        def get *extra_args
+          @mutex.synchronize do
+            # Wait for any backfill to complete, and handle expiration first
+            # because it might change the state.
+            wait_backfill
+            do_expire if should_expire?
+            # Main state handling
+            if @retries.finished?
+              # finished state: return value or error
+              return cached_value
+            elsif !@compute_notify.nil?
+              # computing state: wait for the computing thread to finish then
+              # return its result
+              wait_compute
+              return cached_value
+            else
+              # pending state
+              cur_time = Process.clock_gettime Process::CLOCK_MONOTONIC
+              # waiting for the next retry: return current error
+              raise @error if @expires_at && cur_time < @expires_at
+              # no delay: compute in the current thread
+              enter_compute cur_time
+              # and continue below
+            end
+          end
+
+          # Gets here if we just transitioned from pending to compute
+          perform_compute extra_args
+        end
+
+        ##
+        # This method calls {#get} repeatedly until a final result is available
+        # or retries have exhausted.
+        #
+        # Note: this method spins on {#get}, although honoring any retry delay.
+        # Thus, it is best to call this only if retries are limited or a retry
+        # delay has been configured.
+        #
+        # @param extra_args [Array] extra arguments to pass to the block
+        # @param transient_errors [Array<Class>] An array of exception classes
+        #     that will be treated as transient and will allow await to
+        #     continue retrying. Exceptions omitted from this list will be
+        #     treated as fatal errors and abort the call. Default is
+        #     `[StandardError]`.
+        # @param max_tries [Integer,nil] The maximum number of times this will
+        #     call {#get} before giving up, or nil for a potentially unlimited
+        #     number of attempts. Default is 1.
+        # @param max_time [Numeric,nil] The maximum time in seconds this will
+        #     spend before giving up, or nil (the default) for a potentially
+        #     unlimited timeout.
+        # @param delay_epsilon [Numeric] An extra delay in seconds to ensure
+        #     that retries happen after the retry delay period
+        #
+        # @return [Object] the value
+        # @raise [Exception] if a fatal error happened, or retries have been
+        #     exhausted.
+        #
+        def await *extra_args, transient_errors: nil, max_tries: 1, max_time: nil, delay_epsilon: 0.0001
```

Comment:
Just curious why we're not using retry gem for retrying.

Reply:
Because it's not a simple retry, but one whose conditions and delays are dynamic and tied closely to information from the state of the LazyValue. For example, to determine the delay until the next retry, we need to get the wait time expiration, that is, the next time the computation engine will allow another computation attempt to start.